mkosi/config.py

```diff
@@ -3610,108 +3674,90 @@ def finalize_defaults() -> None:
+    argparser = create_argument_parser(ConfigAction)
+    argparser.parse_args(argv, ParseContext.cli)
+    args = load_args(ParseContext.cli)
 
+    # If --debug was passed, apply it as soon as possible.
     if ARG_DEBUG.get():
         logging.getLogger().setLevel(logging.DEBUG)
 
+    # Do the same for help.
     if args.verb == Verb.help:
-        PagerHelpAction.__call__(None, argparser, namespace)  # type: ignore
+        PagerHelpAction.__call__(None, argparser, ParseContext.cli)  # type: ignore
 
     if not args.verb.needs_config():
         return args, ()
 
-    include = ()
+    # One of the specifiers needs access to the directory so let's make sure it
+    # is available.
+    setattr(ParseContext.config, "directory", args.directory)
 
+    # Parse the global configuration unless the user explicitly asked us not to.
     if args.directory is not None:
-        parse_config_one(Path("."), profiles=True)
+        parse_config_one(Path("."), profiles=True, local=True)
 
-        finalize_default(SETTINGS_LOOKUP_BY_DEST["images"])
-        include = getattr(namespace, "images")
-        immutable_settings.add("Images")
-
-        d: Optional[Path]
-        for d in (Path("mkosi.images"), Path("mkosi.presets")):
-            if Path(d).exists():
-                break
-        else:
-            d = None
-
-        if d:
-            for p in sorted(d.iterdir()):
-                if not p.is_dir() and not p.suffix == ".conf":
-                    continue
-
-                name = p.name.removesuffix(".conf")
-                if not name:
-                    die(f"{p} is not a valid image name")
-
-                ns_copy = copy.deepcopy(namespace)
-                defaults_copy = copy.deepcopy(defaults)
-                parsed_includes_copy = copy.deepcopy(parsed_includes)
-
-                setattr(namespace, "image", name)
-
-                with chdir(p if p.is_dir() else Path.cwd()):
-                    if not parse_config_one(p if p.is_file() else Path(".")):
-                        continue
-
-                finalize_defaults()
-                images += [namespace]
+    # After we've finished parsing the configuration, we'll have values in both
+    # namespaces (ParseContext.cli, ParseContext.config). To be able to parse the values from a
+    # single namespace, we merge the final values of each setting into one namespace.
+    for s in SETTINGS:
+        setattr(ParseContext.config, s.dest, finalize_value(s))
 
-                namespace = ns_copy
-                defaults = defaults_copy
-                parsed_includes = parsed_includes_copy
+    # Load the configuration for the main image.
+    config = load_config(ParseContext.config)
 
-    if not images:
-        setattr(namespace, "image", None)
-        finalize_defaults()
-        images = [namespace]
+    images = []
 
-    append = True
+    if args.directory is not None and (d := Path("mkosi.images")).exists():
```

Comment:
That too would be more readable as
```suggestion
    if args.directory is not None and Path("mkosi.images").exists():
```
and the one usage of `d` below changed as in the other place.